### DISCUSSION TITLE
how to call openmodelica compiler multi thread

### DISCUSSION THREAD

**Discussion by take-study** [OpenModelica/OpenModelica#12478] [Q&A] [answered]
i am writing a web app to use modelica modeling in browser. i use vscode open source lib as my editor gui and plan to use c# to make a http server as `lsp` server. so i wrote some c++ code to wrappe omc just like omshell doing to provide a dll for c#. so that i can built some scripting api command to do something i want. when i call the dll single thread, everything is ok, but when i call the dll in multi thread there are some errors that i can't understand. can someone tell me how to call the omc multi thread or is there a way to made multiple command environment so i can provide independent modeling environment for diffrent browser client

**Comment by arun3688** [comment]
@take-study currently our dll's do not support multi threading, you should create separate process when calling multiple dll's and this will avoid the dead lock, otherwise if you are using multi thread then you have to lock the critical section of the code which i guess is when you do simulation , So i suggest you to use process instead of threads for your web application. 
@adrpo can you comment on this

**Reply by take-study** [reply]
thank for reply, the process looks like a good way to solve this problem but i stll want to know is there a way to create multi environemnt in one process.so that i can load diffrent `.mo` file which has the same model but for diffrent user to operate

**Reply by arun3688** [reply, marked as answer]
@take-study as i said if you use multiple threads on single process, then as i said loading the dll's does not support multithreads and if you want to use threads then you have to lock the critical section of the code for example

```
for i in numofthreads:
    loadDLL("test.dll")
    lock()
    simulate() 
    unlock()
```

**Reply by take-study** [reply]
thank for explain, i will try multi process for my app
